src/Microsoft.OData.Edm/ExtensionMethods/TargetPathHelper.cs

```diff
@@ -0,0 +1,287 @@
+            {
+                return Enumerable.Empty<IEdmElement>();
+            }
+
+            List<IEdmElement> pathSegments = new List<IEdmElement>(targetSegments.Length);
+            pathSegments.Add(container);
+            int index = 1; // index of the targetSegments.
+
+            IEdmVocabularyAnnotatable vocabularyAnnotatable =  HandleEntityContainer(model, pathSegments, container, targetSegments, index, caseInsensitive);
+            index++;
+
+            while (index < targetSegments.Length)
+            {
+                if (vocabularyAnnotatable is IEdmEntityContainerElement containerElement)
+                {
+                    vocabularyAnnotatable = HandleContainerElement(model, pathSegments, containerElement, targetSegments, index, caseInsensitive);
+                }
+                else if (vocabularyAnnotatable is IEdmSchemaType schemaType)
+                {
+                    vocabularyAnnotatable = HandleSchemaType(model, pathSegments, schemaType, targetSegments, index, caseInsensitive);
+                }
+                else if (vocabularyAnnotatable is IEdmProperty property)
+                {
+                    vocabularyAnnotatable = HandleProperty(model, pathSegments, property, targetSegments, index, caseInsensitive);
+                }
+                else
+                {
+                    return Enumerable.Empty<IEdmElement>();
+                }
+
+                index++;
+            }
+
+            return pathSegments;
+        }
+
+        /// <summary>
+        /// Returns an Entity container element from the entity container.
+        /// </summary>
+        /// <param name="model">The edm model.</param>
+        /// <param name="pathSegments">The path segments of already resolved target segments.</param>
+        /// <param name="entityContainer">The entity container.</param>
+        /// <param name="targetSegments">The target segments to resolve.</param>
+        /// <param name="index">The index of the target segments.</param>
+        /// <param name="caseInsensitive">Property name case-insensitive or not.</param>
+        /// <returns>An <see cref="IEdmVocabularyAnnotatable"/> object.</returns>
+        private static IEdmVocabularyAnnotatable HandleEntityContainer(IEdmModel model, List<IEdmElement> pathSegments, IEdmEntityContainer entityContainer, string[] targetSegments, int index, bool caseInsensitive)
+        {
+            IEdmEntitySet entitySet = entityContainer.FindEntitySetExtended(targetSegments[index]);
+
+            if (entitySet != null)
+            {
+                pathSegments.Add(entitySet);
+
+                return entitySet;
+            }
+
+            IEdmSingleton singleton = entityContainer.FindSingletonExtended(targetSegments[index]);
+
+            if (singleton != null)
+            {
+                pathSegments.Add(singleton);
+
+                return singleton;
+            }
+
+            return null;

```

Comment:
For these methods we return null when we do not get the segments. Why can't we throw exceptions.

Reply:
Validations happen when reading a csdl. However the scope of this PR is for segments count 3+. That scenario is already in the existing logic.
https://github.com/OData/odata.net/blob/0209fb9b8d044728fe90d18f22f2bdb361268e1d/src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsVocabularyAnnotation.cs#L361